File: backend/app/api/live_sessions.py
from __future__ import annotations

import logging
from typing import List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Header, HTTPException, Query, status
from pydantic import ValidationError
from app.db.engine import get_session_factory
from app.db.models import User
from app.repositories.commerce_repository import CommerceRepository
from app.services.auth_service import get_optional_user
from app.services.host_lease import verify_host_token
from app.services.live_room_types import room_type_requires_commerce

from app.schemas.live_sessions import (
    CreateLiveRoomRequest,
    CreateLiveRoomResponse,
    HostHeartbeatRequest,
    LiveRoomResponse,
    LiveSessionResponse,
    ModerationViolationRequest,
    ReclaimHostRequest,
    StartLiveSessionRequest,
)
from app.services.live_session_moderation import (
    HostLeaseError,
    LiveRoomValidationError,
    LiveSessionModerationError,
    apply_moderation_violation,
    create_live_room,
    end_live_session,
    get_active_live_session,
    get_live_session,
    host_heartbeat,
    list_active_live_sessions,
    reclaim_host,
    start_live_session,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/live-sessions", tags=["live-sessions"])


def _authorize_session_control(
    session_id: UUID,
    user: Optional[User],
    host_token: Optional[str],
) -> object:
    session = get_live_session(session_id)
    if session is None:
        raise HTTPException(status_code=404, detail="Session was not found.")
    if user is not None and getattr(session, "seller_user_id", None) == user.id:
        return session
    metadata = dict(getattr(session, "metadata_json", None) or {})
    if host_token and verify_host_token(metadata, host_token):
        return session
    if user is None and not host_token:
        raise HTTPException(status_code=401, detail="Host authorization required.")
    raise HTTPException(status_code=403, detail="Caller cannot control this session.")


@router.get("", response_model=List[LiveRoomResponse], include_in_schema=True)
@router.get("/", response_model=List[LiveRoomResponse], include_in_schema=False)
def api_list_live_sessions(
    status_filter: Optional[str] = Query(default="active", alias="status"),
) -> List[LiveRoomResponse]:
    normalized = (status_filter or "active").strip().lower()
    if normalized != "active":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Only status=active is supported in this MVP.",
        )
    sessions = list_active_live_sessions()
    return [LiveRoomResponse.from_session(session) for session in sessions]


@router.post(
    "",
    response_model=CreateLiveRoomResponse,
    status_code=status.HTTP_201_CREATED,
    include_in_schema=True,
)
@router.post(
    "/",
    response_model=CreateLiveRoomResponse,
    status_code=status.HTTP_201_CREATED,
    include_in_schema=False,
)
def api_create_live_room(
    request: CreateLiveRoomRequest,
    user: Optional[User] = Depends(get_optional_user),
) -> CreateLiveRoomResponse:
    try:
        shop_id = None
        if room_type_requires_commerce(request.room_type):
            if user is None:
                raise HTTPException(
                    status_code=401,
                    detail="Sign in before creating a commerce room.",
                )
            with get_session_factory() as db_session:
                repo = CommerceRepository(db_session)
                shop = repo.shop_for_owner(user.id)
                if shop is None:
                    raise HTTPException(
                        status_code=409,
                        detail="Create a shop before a commerce room.",
                    )
                if not request.product_ids:
                    raise HTTPException(
                        status_code=422,
                        detail="Select at least one product for a commerce room.",
                    )
                for product_id in set(request.product_ids):
                    product = repo.product(product_id, include_inactive=True)
                    if (
                        product is None
                        or product.shop_id != shop.id
                        or not product.is_active
                    ):
                        raise HTTPException(
                            status_code=409,
                            detail="All selected products must be active and belong to your shop.",
                        )
                shop_id = shop.id
        session, host_token = create_live_room(
            request.name,
            request.room_type,
            seller_user_id=user.id if user else None,
            shop_id=shop_id,
        )
    except LiveRoomValidationError as error:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={"code": error.code, "message": error.message},
        ) from error
    except ValidationError as error:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error.errors(),
        ) from error
    base = LiveRoomResponse.from_session(session)
    return CreateLiveRoomResponse(**base.model_dump(), host_resume_token=host_token)


@router.post("/start", response_model=LiveSessionResponse)
def api_start_live_session(
    request: StartLiveSessionRequest,
    user: Optional[User] = Depends(get_optional_user),
) -> LiveSessionResponse:
    shop_id = None
    if user is not None:
        with get_session_factory() as db_session:
            shop = CommerceRepository(db_session).shop_for_owner(user.id)
            if shop is None:
                raise HTTPException(409, detail="Create a shop before starting a seller room.")
            shop_id = shop.id
    session = start_live_session(
        request.room_id,
        seller_user_id=user.id if user else None,
        shop_id=shop_id,
    )
    return LiveSessionResponse.from_session(session)


@router.get("/by-room/{room_id}/current", response_model=LiveSessionResponse)
def api_get_current_live_session(room_id: str) -> LiveSessionResponse:
    session = get_active_live_session(room_id)
    if session is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"No active live session for room '{room_id}'.",
        )
    return LiveSessionResponse.from_session(session)


@router.post("/by-room/{room_id}/reclaim-host", response_model=LiveSessionResponse)
def api_reclaim_host(room_id: str, request: ReclaimHostRequest) -> LiveSessionResponse:
    try:
        session = reclaim_host(room_id, host_token=request.host_token)
    except HostLeaseError as error:
        status_code = (
            status.HTTP_404_NOT_FOUND
            if error.code in {"session_not_found", "host_lease_expired"}
            else status.HTTP_403_FORBIDDEN
        )
        raise HTTPException(
            status_code=status_code,
            detail={"code": error.code, "message": error.message},
        ) from error
    return LiveSessionResponse.from_session(session, is_host=True)


@router.post("/{session_id}/host-heartbeat", response_model=LiveSessionResponse)
def api_host_heartbeat(session_id: UUID, request: HostHeartbeatRequest) -> LiveSessionResponse:
    try:
        session = host_heartbeat(
            session_id,
            host_token=request.host_token,
            media_live=request.media_live,
        )
    except HostLeaseError as error:
        status_code = (
            status.HTTP_404_NOT_FOUND
            if error.code in {"session_not_found", "host_lease_expired"}
            else status.HTTP_403_FORBIDDEN
        )
        raise HTTPException(
            status_code=status_code,
            detail={"code": error.code, "message": error.message},
        ) from error
    return LiveSessionResponse.from_session(session, is_host=True)


@router.post("/{session_id}/end", response_model=LiveSessionResponse)
async def api_end_live_session(
    session_id: UUID,
    user: Optional[User] = Depends(get_optional_user),
    host_token: Optional[str] = Header(default=None, alias="X-Host-Token"),
) -> LiveSessionResponse:
    _authorize_session_control(session_id, user, host_token)
    try:
        session, already_ended = end_live_session(session_id, ended_reason="host_stopped")
    except LiveSessionModerationError as error:
        if error.code == "session_not_found":
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail={"code": error.code, "message": error.message},
            ) from error
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={"code": error.code, "message": error.message},
        ) from error

    if not already_ended:
        await _broadcast_session_ended(session)

    return LiveSessionResponse.from_session(session, already_ended=already_ended)


@router.post("/{session_id}/moderation-violations", response_model=LiveSessionResponse)
async def api_report_moderation_violation(
    session_id: UUID,
    request: ModerationViolationRequest,
    user: Optional[User] = Depends(get_optional_user),
    host_token: Optional[str] = Header(default=None, alias="X-Host-Token"),
) -> LiveSessionResponse:
    _authorize_session_control(session_id, user, host_token)
    try:
        session, already_ended = apply_moderation_violation(session_id, request)
    except LiveSessionModerationError as error:
        if error.code == "session_not_found":
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail={"code": error.code, "message": error.message},
            ) from error
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={"code": error.code, "message": error.message},
        ) from error

    if not already_ended:
        await _broadcast_session_ended(session)

    return LiveSessionResponse.from_session(session, already_ended=already_ended)


async def _broadcast_session_ended(session: object) -> None:
    from app.api.chat import chat_manager
    from app.services.memory_live_sessions import VISUAL_MODERATION_ENDED_REASON

    metadata = getattr(session, "metadata_json", None) or {}
    room_id = getattr(session, "room_id")
    session_id = getattr(session, "id")
    ended_at = getattr(session, "ended_at", None)
    await chat_manager.broadcast_event(
        room_id,
        {
            "type": "live_session_ended",
            "reason": metadata.get("ended_reason") or VISUAL_MODERATION_ENDED_REASON,
            "session_id": str(session_id),
            "room_id": room_id,
            "ended_at": ended_at.isoformat() if ended_at is not None else None,
        },
    )

File: frontend/src/api/liveSessions.ts
import { getApiBaseUrl } from "./config";
import { authHeaders } from "./auth";

export type LiveSessionStatus = "active" | "ended";

export type LiveSession = {
  id: string;
  room_id: string;
  status: LiveSessionStatus;
  started_at: string;
  ended_at: string | null;
  ended_reason: string | null;
  metadata: Record<string, unknown>;
  already_ended?: boolean;
  name?: string | null;
  room_type?: string | null;
  host_present?: boolean;
  host_recoverable?: boolean;
  host_lease_expires_at?: string | null;
  media_live?: boolean;
  is_host?: boolean;
  grace_remaining_seconds?: number | null;
  seller_user_id?: string | null;
  shop_id?: string | null;
};

export type LiveRoom = {
  id: string;
  room_id: string;
  name: string;
  room_type: string;
  status: LiveSessionStatus;
  started_at: string;
  ended_at: string | null;
  metadata: Record<string, unknown>;
  host_present?: boolean;
  host_recoverable?: boolean;
  host_lease_expires_at?: string | null;
  media_live?: boolean;
  host_resume_token?: string;
  seller_user_id?: string | null;
  shop_id?: string | null;
};

export type HostHeartbeatPayload = {
  host_token: string;
  media_live?: boolean;
};

export type CreateLiveRoomPayload = {
  name: string;
  room_type: string;
  product_ids?: string[];
};

export type ModerationViolationPayload = {
  code: "sharp_object_detected";
  label: "knife" | "scissors";
  confidence: number;
  evidence_count: number;
  window_ms: number;
  detected_at: string;
};

export async function listActiveLiveRooms(): Promise<LiveRoom[]> {
  const response = await fetch(`${getApiBaseUrl()}/api/live-sessions?status=active`);
  if (!response.ok) {
    throw new Error(`Failed to load live rooms (${response.status}).`);
  }
  return (await response.json()) as LiveRoom[];
}

export async function createLiveRoom(payload: CreateLiveRoomPayload): Promise<LiveRoom> {
  const response = await fetch(`${getApiBaseUrl()}/api/live-sessions`, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    let detail = `Failed to create live room (${response.status}).`;
    try {
      const body = (await response.json()) as { detail?: unknown };
      if (typeof body.detail === "string") {
        detail = body.detail;
      } else if (
        body.detail &&
        typeof body.detail === "object" &&
        "message" in body.detail &&
        typeof (body.detail as { message: unknown }).message === "string"
      ) {
        detail = (body.detail as { message: string }).message;
      }
    } catch {
      // keep default
    }
    throw new Error(detail);
  }
  return (await response.json()) as LiveRoom;
}

export async function sendHostHeartbeat(
  sessionId: string,
  payload: HostHeartbeatPayload,
): Promise<LiveSession> {
  const response = await fetch(
    `${getApiBaseUrl()}/api/live-sessions/${encodeURIComponent(sessionId)}/host-heartbeat`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    },
  );
  if (!response.ok) {
    throw new Error(`Host heartbeat failed (${response.status}).`);
  }
  return (await response.json()) as LiveSession;
}

export async function reclaimHost(roomId: string, hostToken: string): Promise<LiveSession> {
  const response = await fetch(
    `${getApiBaseUrl()}/api/live-sessions/by-room/${encodeURIComponent(roomId)}/reclaim-host`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ host_token: hostToken }),
    },
  );
  if (!response.ok) {
    throw new Error(`Failed to reclaim host (${response.status}).`);
  }
  return (await response.json()) as LiveSession;
}

export async function startLiveSession(roomId: string): Promise<LiveSession> {
  const response = await fetch(`${getApiBaseUrl()}/api/live-sessions/start`, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify({ room_id: roomId }),
  });
  if (!response.ok) {
    throw new Error(`Failed to start live session (${response.status}).`);
  }
  return (await response.json()) as LiveSession;
}

export async function endLiveSession(
  sessionId: string,
  hostToken?: string | null,
): Promise<LiveSession> {
  const response = await fetch(
    `${getApiBaseUrl()}/api/live-sessions/${encodeURIComponent(sessionId)}/end`,
    {
      method: "POST",
      headers: {
        ...authHeaders(),
        ...(hostToken ? { "X-Host-Token": hostToken } : {}),
      },
    },
  );
  if (!response.ok) {
    throw new Error(`Failed to end live session (${response.status}).`);
  }
  return (await response.json()) as LiveSession;
}

export async function reportModerationViolation(
  sessionId: string,
  payload: ModerationViolationPayload,
  hostToken?: string | null,
): Promise<LiveSession> {
  const response = await fetch(
    `${getApiBaseUrl()}/api/live-sessions/${encodeURIComponent(sessionId)}/moderation-violations`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...authHeaders(),
        ...(hostToken ? { "X-Host-Token": hostToken } : {}),
      },
      body: JSON.stringify(payload),
    },
  );
  if (!response.ok) {
    throw new Error(`Failed to report moderation violation (${response.status}).`);
  }
  return (await response.json()) as LiveSession;
}

export async function getCurrentLiveSession(roomId: string): Promise<LiveSession | null> {
  const response = await fetch(
    `${getApiBaseUrl()}/api/live-sessions/by-room/${encodeURIComponent(roomId)}/current`,
  );
  if (response.status === 404) {
    return null;
  }
  if (!response.ok) {
    throw new Error(`Failed to load current live session (${response.status}).`);
  }
  return (await response.json()) as LiveSession;
}
